### ISSUE TITLE
eSpeak Out of Memory issue

### ISSUE THREAD

**Issue by cvl01** [echogarden-project/echogarden#64]
When aligning audio with a transcript, for longer audio files (I have audio's between 1h:00m - 1h:30m long), I get the following error:

```
Synthesize recognized transcript with eSpeak.. Aborted(OOM)
Error: Aborted(OOM). Build with -sASSERTIONS for more info.
```

Another series of audio files shorter than 1h does work succesfully!

I looked through the code but could not find out if any memory restrictions are in place. Do you have an idea how to solve this problem?

**Comment by rotemdan** [comment]
I've never seen this particular error in any Node.js program I ran.

Maybe you're using a special version of Node.js? What OS do you use? 

The alignment is tested to up to 6 hours. It doesn't look like a memory issue.

The call it's doing is to code that's been compiled from C, using Emscripten to JavaScript. The error may occur during that execution, but I can't identify what it's saying.

I would go and ask an LLM about this particular error to try to get some clues.

**Comment by rotemdan** [comment]
Maybe OOM means "out of memory". I've seen "out-of-memory" issues but I don't remember they had this particular error.

**Comment by cvl01** [comment]
I am using node v18.20.4, on Windows.

**Comment by rotemdan** [comment]
The official minimum version of the package is `18.16.0`. Although it's technically supported, it's still considered an old version now.

I've previously verified that versions of Node.js after `22.0.0` have less memory limits. I think for previous version there was a limit to an ArrayBuffer that was about 2GB or 4GB, but after `22.0.0` it seemed to have been removed since I could allocate much larger ones.

Try [v22.5.1](https://nodejs.org/dist/v22.5.1/node-v22.5.1-x64.msi) and see if that changes anything.

**Comment by cvl01** [comment]
I have tried v22.5.1 but so far no luck, still getting the same error on longer audio files.

I'm now planning to test it from a project, to see if running it from a typescript project gives more detailed error messages than via the CLI.

**Comment by rotemdan** [comment]
WASM memory space is limited to either 2GB or 4GB. 64 bit WebAssembly is still a proposal, pointers are always 32 bit until that happens. Although eSpeak is compiled to JavaScript, it's possible that the limitation exists due to compatibility reasons in the Emscripten compiler.

Seems like eSpeak is trying to allocate a lot of memory. It could be for a very large audio file (more than 2GB or 4GB) or for something else.

I did actually investigate alignment of very long audio file, a few months ago, but never got out of memory issues with eSpeak actually. I got it with other things like the sample rate converter (fixed to use chunking so no problem now), the wave format encoder and decoder, and FFMpeg.

Some important relevant information:

What is the duration of the audio file you're using?
What is the length of the transcript you're using?

It could be that in your case the transcript is extremely long, so synthesizing it produces an audio file that goes beyond the memory space, either that or some other memory structure eSpeak uses becomes too large.

**Comment by cvl01** [comment]
The audio file is 01:15 (75 mins). The plaintext reference transcript is 63KB. The whisper generated transcript is probably a bit longer, since I see some occasional hallucination in the CLI output. 

I am using align with dtw-ra and whisper as recognition model. Synthesizing the ground-truth works well, the tool errors on the next step. 

```console
[01:15:45.420 --> 01:15:47.420]   *gelach*

407087.1ms
Load tokenizer data.. 81.6ms
Total recognition time: 407953.4ms

Synthesize ground-truth transcript with eSpeak.. 13557.2ms
Synthesize recognized transcript with eSpeak.. Aborted(OOM)
Error: Aborted(OOM). Build with -sASSERTIONS for more info.
PS C:\Users\luik001c>
````

**Comment by rotemdan** [comment]
The extra information really helps.

`dtw-ra` synthesizes both the ground-truth transcript and the recognized transcript produced by the recognition model. 

If the recognized contains lots of repetition it is possible it would become so long that its synthesized version would require a very large amount of memory. It's also possible the number of characters in it is beyond what eSpeak can handle in the memory space it has - not necessarily due to the audio being too large.

The audio eSpeak natively produces is 16 bit 22050 Hz mono, which does makes it slightly larger than the typical 16000 Hz used internally. Maybe I could ask it to produce its output 16000 Hz instead to save a bit of memory, but that wouldn't necessarily dramatically improve the memory size of the audio, possibly by 37% but that's only in case it natively produces the audio in a 16000 Hz sample rate (which is reasonable to assume if it's efficient in that - so maybe).

**Comment by rotemdan** [comment]
Wait, the synthesis is already done in chunks - I believe sentence by sentence, so memory shouldn't be an issue. 

I'm not sure what's going on. I'll need to investigate further, but that may not happen today.

**Comment by rotemdan** [comment]
I looked at the code. It seems like for `dtw-ra` I synthesize the recognized transcript all in one go. That's not the way I do it anywhere else, I believe. 

I could fix it by doing this in chunks in that case.

**Comment by cvl01** [comment]
First, Thanks for providing such detailed and excellent support!

This is the output running it in a node project:
```
C:\Users\luik001c\scoop\apps\nodejs-lts\current\node.exe .\out\test-copy.js
Processing Blok ProRail 20231004 GELUIDSOPNAME.wav...
Aborted(OOM)
Uncaught Error Error: Aborted(OOM). Build with -sASSERTIONS for more info.
    at abort (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:23:2925)
    at abortOnCannotGrowMemory (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:23:59631)
    at _emscripten_resize_heap (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:23:59750)
    at Wb (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:14:30897)
    at Qa (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:13:210230)
    at We (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:14:32769)
    at Me (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:14:1839)
    at wb (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:14:22024)
    at dg (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:14:26018)
    at eSpeakNGWorker.set_voice.eSpeakNGWorker.set_voice (c:\Users\luik001c\echogarden-github\node_modules\@echogarden\espeak-ng-emscripten\espeak-ng.cjs:23:97968)
    at setVoice (c:\Users\luik001c\echogarden-github\dist\synthesis\EspeakTTS.js:377:14)
Process exited with code 1
```

**Comment by cvl01** [comment]
I'll look at the code an see if I can change it to chunking myself, I have the environment setup but am ofcourse not familiar with the codebase, otherwise I'll wait until you'll have time to look at it. 

I can also split my files in half but since I am planning to use it more often getting this working with the original length files/transcripts would be nice!

I very much like the library and its features, thanks a lot for that!

**Comment by rotemdan** [comment]
Splitting may not the best thing to do. 

Alignment is usually better done over the entire audio and transcript.

I don't think you should try to work on it yourself. There are some very subtle things it's doing in how it passes the text to eSpeak, like producing "fragments", etc.

The code is in `SpeechAlignment.ts`:

In `alignUsingDtwWithRecognition`:
```ts
	// Synthesize the recognized transcript and get its timeline
	logger.start("Synthesize recognized transcript with eSpeak")
	const recognizedWords = recognitionTimeline.map(entry => entry.text)

	const {
		rawAudio: synthesizedRecognizedTranscriptRawAudio,
		timeline: synthesizedRecognitionTimeline
	} = await createAlignmentReferenceUsingEspeakForFragments(recognizedWords, espeakOptions)
```

And `createAlignmentReferenceUsingEspeakForFragments` is:
```ts
export async function createAlignmentReferenceUsingEspeakForFragments(fragments: string[], espeakOptions: EspeakOptions) {
	const progressLogger = new Logger()

	progressLogger.start("Load espeak module")
	const Espeak = await import("../synthesis/EspeakTTS.js")

	progressLogger.start("Synthesize alignment reference with eSpeak")

	const result = await Espeak.synthesizeFragments(fragments, espeakOptions)

	result.timeline = result.timeline.flatMap(clause => clause.timeline!)

	for (const wordEntry of result.timeline) {
		wordEntry.timeline = wordEntry.timeline!.flatMap(tokenEntry => tokenEntry.timeline!)
	}

	progressLogger.end()

	return result
}
```

That is the only call to `createAlignmentReferenceUsingEspeakForFragments` in the entire codebase. 

I changed all the other calls, during alignment to use the full synthesis API (`synthesize`).

The reason `synthesize` is not used here is that it because `createAlignmentReferenceUsingEspeakForFragments` preserves the word segmentation provided by the recognizer so that later the word events would match with the original timing. `synthesize` only accepts transcripts. You can't give it a list of already-segmented words.

`synthesize` also does a lot of things that are not necessary here, like segmenting to sentences and segments, normalization and other pre-processing. In general it would be slower, but much more memory efficient.

The change would require to split the fragments to chunks and synthesize each chunk separately, and then merge the synthesized and their timelines, and make corrections to the timing information. It's not a trivial change to make and there's no reason for someone to spend hours on this if they are not familiar with the library and similar code used in other places.

I never really considered doing this, maybe since the `dtw-ra` used to be considered "experiential" and slow, so I didn't expect people would use it for such long inputs.

**Comment by cvl01** [comment]
Sure, I'll wait for you to make changes.
I am using dtw-ra since so far it seems more accurate than dtw on my audio, handling difficult audio portions better.

**Comment by rotemdan** [comment]
I added a fix for this issue on the new `1.6.0` release.

The source diff is [here](https://github.com/echogarden-project/echogarden/commit/7d64ec10840f4806edb7e79e0f43652de1a6b081).

Let me know if you encounter any other issues.

**Comment by rotemdan** [comment]
Turns out that didn't fully fix the issue.

It took many hours of effort but I eventually identified the cause.

I eventually could reproduce the out-of-memory problem by simply setting the espeak voice over and over again, in a loop.

It was in how `espeak-ng` internally allocates new memory whenever a voice is set.

In `src/libespeak-ng/speech.c`:
```c
espeak_ng_STATUS DoVoiceChange(voice_t *v)
{
	// allocate memory for a copy of the voice data, and free it in wavegenfill()
	voice_t *v2;
	if ((v2 = (voice_t *)malloc(sizeof(voice_t))) == NULL)
		return ENOMEM;
	memcpy(v2, v, sizeof(voice_t));
	wcmdq[wcmdq_tail][0] = WCMD_VOICE;
	wcmdq[wcmdq_tail][2] = (intptr_t)v2;
	WcmdqInc();
	return ENS_OK;
}
```

The problem is that it expects the allocated memory to be released in `wavegenfill`, but some utility methods like `text_to_phonemes`, which is called very often in my code, set the voice internally but don't actually generate any speech, causing a kind of memory leak / overload.

The fix was to disable any non-essential calls to `set-voice` in `espeakng_glue.cpp` (part of the emscripten build of `espeak-ng`), and in `EspeakTTS.ts`.

I also enabled `ALLOW_MEMORY_GROWTH` in the new `espeak-ng-emscripten` build, so even if there's a memory leak somewhere else, it's less likely to cause a crash, unless it reaches the maximum of 2GB / 4GB of memory. It may not be fully necessary, but it's a kind of an added safeguard.
